blockchain/v2/reactor.go

```diff
@@ -1,118 +1,521 @@
+}
+
+type blockVerifier interface {
+	VerifyCommit(chainID string, blockID types.BlockID, height int64, commit *types.Commit) error
 }
 
-func NewReactor(bufferSize int) *Reactor {
-	return &Reactor{
+// XXX: unify naming in this package around tdState
+// XXX: V1 stores a copy of state as initialState, which is never mutated. Is that nessesary?
+func newReactor(state state.State, store blockStore, reporter behaviour.Reporter,
+	blockApplier blockApplier, bufferSize int) *BlockchainReactor {
+	scheduler := newScheduler(state.LastBlockHeight, time.Now())
+	pContext := newProcessorContext(store, blockApplier, state)
+	// TODO: Fix naming to just newProcesssor
+	// newPcState requires a processorContext
+	processor := newPcState(pContext)
+
+	return &BlockchainReactor{
 		events:    make(chan Event, bufferSize),
 		stopDemux: make(chan struct{}),
-		scheduler: newRoutine("scheduler", schedulerHandle, bufferSize),
-		processor: newRoutine("processor", processorHandle, bufferSize),
-		ticker:    time.NewTicker(1 * time.Second),
+		scheduler: newRoutine("scheduler", scheduler.handle, bufferSize),
+		processor: newRoutine("processor", processor.handle, bufferSize),
+		store:     store,
+		reporter:  reporter,
 		logger:    log.NewNopLogger(),
 	}
 }
 
-// nolint:unused
-func (r *Reactor) setLogger(logger log.Logger) {
+// NewBlockchainReactor creates a new reactor instance.
+func NewBlockchainReactor(state state.State, blockApplier blockApplier, store blockStore, fastSync bool) *BlockchainReactor {
+	reporter := behaviour.NewMockReporter()
+	return newReactor(state, store, reporter, blockApplier, 1000)
+}
+
+// SetSwitch implements Reactor interface.
+func (r *BlockchainReactor) SetSwitch(sw *p2p.Switch) {
+	if sw == nil {
+		panic("Initalized BlockchainReactor with a nil switch")
+	}
+
+	r.Switch = sw
+	r.io = newSwitchIo(sw)
+}
+
+func (r *BlockchainReactor) setMaxPeerHeight(height int64) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	if height > r.maxPeerHeight {
+		r.maxPeerHeight = height
+	}
+}
+
+func (r *BlockchainReactor) setSyncHeight(height int64) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	r.syncHeight = height
+}
+
+// SyncHeight returns the height to which the BlockchainReactor has synced.
+func (r *BlockchainReactor) SyncHeight() int64 {
+	r.mtx.RLock()
+	defer r.mtx.RUnlock()
+	return r.syncHeight
+}
+
+// SetLogger sets the logger of the reactor.
+func (r *BlockchainReactor) SetLogger(logger log.Logger) {
 	r.logger = logger
 	r.scheduler.setLogger(logger)
 	r.processor.setLogger(logger)
 }
 
-func (r *Reactor) Start() {
+// Start implements cmn.Service interface
+func (r *BlockchainReactor) Start() error {
+	r.reporter = behaviour.NewSwitcReporter(r.BaseReactor.Switch)
 	go r.scheduler.start()
 	go r.processor.start()
 	go r.demux()
+	return nil
+}
 
-	<-r.scheduler.ready()
-	<-r.processor.ready()
+// reactor generated ticker events:
+// ticker for cleaning peers
+type rTryPrunePeer struct {
```

Comment:
why do we need `r` prefix? because events are used outside of this file?

Reply:
We prefix the event with `r` to indicate that this is a reactor event.

Reply:
could this be added to the `doc.go` file so someone doesn't remove it in the future